### DISCUSSION TITLE
[Feature] Support Android's new cloud media provider for stock photo picker

### DISCUSSION THREAD

**Discussion by bdonvr** [immich-app/immich#7101] [Feature Request]
### The feature

https://android-developers.googleblog.com/2024/02/cloud-photos-now-available-in-android-photo-picker.html

It says this is currently in pilot stage for apps nominated by OEMs, but it might be good to get the feature ready for wider rollout.

More info: https://developer.android.com/guide/topics/providers/cloud-media-provider

### Platform

- [ ] Server
- [ ] Web
- [X] Mobile

**Comment by Alex1607** [comment]
Is there any update on this and as if this will be included soonish? 
I think this would be an awesome addition, and personally it would help me out a lot as I then don't have to first switch apps to download an image first or just have a copy in google photos all the time.

**Comment by dannykorpan** [comment]
+1

**Comment by clari7744** [comment]
If it would be possible to add full search capabilities including "search by face", this would be great - especially for finding contact photos; it's much easier when I can just get all that person's photos.

**Comment by NolandTech** [comment]
![Screenshot_20250408_201614_Media picker.jpg](https://github.com/user-attachments/assets/d5c238c1-b54a-4af4-a2b3-3c7806e3989f)

When?

**Comment by Simonxdd** [comment]
This would be great to have in the future. I find it very convenient with Google Photos.

**Comment by bl4zee1g** [comment]
Any update?

**Reply by bo0tzz** [reply]
https://justinmayer.com/posts/any-updates/

**Comment by quittung** [comment]
Since a few people have asked: this feature (CloudMediaProvider) is still restricted by Google to a small number of apps that are pre-approved by the device manufacturer (OEM). Immich isn’t on that list, so even if support were added, Android wouldn’t show it in the system photo picker.

It’s part of a closed pilot program, and as of now there’s no indication from Google that they plan to open it up more broadly.

**Reply by Alex1607** [reply]
> Since a few people have asked: this feature (CloudMediaProvider) is still restricted by Google to a small number of apps that are pre-approved by the device manufacturer (OEM). Immich isn’t on that list, so even if support were added, Android wouldn’t show it in the system photo picker.
> 
> It’s part of a closed pilot program, and as of now there’s no indication from Google that they plan to open it up more broadly.

I mean, Immich has FUTO as a sponsor. Maybe they can get in contact with an OEM like Fairphone which have an interest overlap which they can partner on. Fairphone already ships phones with an e/os so they clearly are pro privacy.

**Reply by Dreaming-Codes** [reply]
I’d love to work on this feature. If the Immich team confirms they can get the OEM provider issue sorted out with FUTU, I’ll start working on it.

**Reply by Dreaming-Codes** [reply]
I may have found a workaround working on it...

**Reply by Dreaming-Codes** [reply]
<img width="591" height="1280" alt="image" src="https://github.com/user-attachments/assets/e0066be0-10f9-4863-afd7-9da89763e823" />
👀

**Reply by Dreaming-Codes** [reply]
So... it works...

The code quality is nowhere near PR-ready yet. I’ll work on it a bit more and then open a PR.

Because of that Google limitation, this workaround requires an ADB command to be run. However, that part is automated thanks to Shizuku, so as long as you're connected to Wi-Fi, which is required for [Shizuku](https://github.com/thedjchi/Shizuku) to work, it will automatically run the needed ADB command after a reboot. This is only needed on boot.

Hopefully, FUTU can get a smartphone manufacturer to add Immich to the allowed apps so we can avoid this hack.

**Reply by Dreaming-Codes** [reply]
Apparently I was wrong it also persists across reboots so it's a one time command.
I'll create a PR asap

**Reply by Dreaming-Codes** [reply]
PR is live #26779

**Comment by Emeraldiaa** [comment]
I'd really love to see this become a thing.
Somewhat new to immich and this is one golden feature that I hope happens. 

To all the people working on this love yall. Fingers crossed

**Comment by timo1302** [comment]
Hello,

alternatively it is possible with the "storage acceess framework" SAF.

E.g. it is also available for Nextcloud:

![Screenshot_20250713-112740~2.png](https://github.com/user-attachments/assets/95876ffd-15d5-4213-9777-5c49470ee169)


Details: https://developer.android.com/guide/topics/providers/document-provider?hl=en

With this feature other Apps at the android phone can access photos inside immich native. Without to download photos from immich App manually before

I think there is no OEM restriction by google. If the cloud media provider integration is not possible.

**Reply by funigui** [reply]
This would help, its probably the only negative to Immich

**Reply by griffi-gh** [reply]
im kinda surprised that one isn't implemented yet tbh

**Reply by kingp0dd** [reply]
I love this feature from Google photos. Especially when needing to upload old photos

**Reply by timo1302** [reply]
I hope too that it will be implemented soon.

**Comment by Dreaming-Codes** [comment]
Implemented in #26779

**Reply by McBaumwolle** [reply]
As they are [not willing](https://github.com/immich-app/immich/pull/26779#issuecomment-4024009593) to implement this fantastic and easy solution, one way would be to mirror this repository and implement the changes there, however this sounds like a lot of unnecessary work (long term). :/

**Reply by Kuuchuu** [reply]
I really don't understand why we can't have a Shizuku toggle in the app's already existing advanced settings with a link to the Immich documentation on what ADB commands the toggle runs. Seems like a no-brainer. There is a solution for a major app inconvenience right here but we instead have to wait on Google to change their mind on this. Would love a mirror with these changes, this is the main thing keeping my family from using Immich.

**Reply by NicholasFlamy** [reply]
> I really don't understand why we can't have a Shizuku toggle in the app's already existing advanced settings with a link to the Immich documentation on what ADB commands the toggle runs. Seems like a no-brainer. There is a solution for a major app inconvenience right here but we instead have to wait on Google to change their mind on this. Would love a mirror with these changes, this is the main thing keeping my family from using Immich.

Afaik, currently the release on the Play Store matches the release on GitHub. Google doesn't like ADB/Shizuku so that might affect Play Store availability and right now Immich is competing with the photos app that must not be named, not Google/Google Play itself.

Keep in mind, what I am saying is just my opinion based on publicly available information. I don't speak for Immich.

**Reply by matejkramny** [reply]
Yes but it could be implemented and not mentioned therefore not violating play rules

**Reply by NicholasFlamy** [reply]
> Yes but it could be implemented and not mentioned therefore not violating play rules

Yes, but how would one know if it breaks the app on certain OEMs or after an Android update? Too risky, and not worth developer time to keep it in there unless it's being used. They're busy with other features and issues.

**Reply by Kuuchuu** [reply]
> > I really don't understand why we can't have a Shizuku toggle in the app's already existing advanced settings with a link to the Immich documentation on what ADB commands the toggle runs. Seems like a no-brainer. There is a solution for a major app inconvenience right here but we instead have to wait on Google to change their mind on this. Would love a mirror with these changes, this is the main thing keeping my family from using Immich.
> 
> Afaik, currently the release on the Play Store matches the release on GitHub. Google doesn't like ADB/Shizuku so that might affect Play Store availability and right now Immich is competing with the photos app that must not be named, not Google/Google Play itself.
> 
> Keep in mind, what I am saying is just my opinion based on publicly available information. I don't speak for Immich.

If the issue is Shizuku features on the play store app then it would have been nice for the devs to state that, rather than that they felt uncomfortable about the proposed solution and locking it from further discussion. I would have been disappointed, but it would be a lot more understandable.

If that truly is the case, could this solution be implemented as an unofficial companion app that adds the functionality?

**Reply by NicholasFlamy** [reply]
> If the issue is Shizuku features on the play store app then it would have been nice for the devs to state that

Let me be clear, this is my suspicion based on what's happened to other apps. Not a statement from the devs.

> If that truly is the case, could this solution be implemented as an unofficial companion app that adds the functionality?

I suspect they don't want to chase workarounds. At all. Once again, priorities.

**Reply by Dreaming-Codes** [reply]
@zackpollard My PR also included the document provider implementation that does not need approval by google and it already works so I think it should be reopened. I think it also make sense to have both merged togheter as they share most of the code.

**Reply by Kuuchuu** [reply]
> > If the issue is Shizuku features on the play store app then it would have been nice for the devs to state that
> 
> Let me be clear, this is my suspicion based on what's happened to other apps. Not a statement from the devs.
> 
> > If that truly is the case, could this solution be implemented as an unofficial companion app that adds the functionality?
> 
> I suspect they don't want to chase workarounds. At all. Once again, priorities.

Sorry, wasn't requesting a companion app from the Immich devs, more from Dreaming-Codes on if it was possible for this solution to be separated from the actual app.  I'm not as familiar with Kotlin & Dart as I'd like to be at the moment so I was curious if that would be possible. I'm assuming it wouldn't be unless some sort of plugin functionality was implemented into Immich, which I know wouldn't be happening.

(wasn't requesting anything, just curious on what is possible)

**Reply by NicholasFlamy** [reply]
> My PR also included the document provider implementation that does not need approval by google and it already works so I think it should be reopened

Split it out into a new branch and open a new PR just for that.
(No mention of ADB in the new branch/PR.)
That one solves #1641

**Reply by Dreaming-Codes** [reply]
> > My PR also included the document provider implementation that does not need approval by google and it already works so I think it should be reopened
> 
> Split it out into a new branch and open a new PR just for that. (No mention of ADB in the new branch/PR.) That one solves #1641

Done #26802

**Reply by Dreaming-Codes** [reply]
> > > If the issue is Shizuku features on the play store app then it would have been nice for the devs to state that
> > 
> > 
> > Let me be clear, this is my suspicion based on what's happened to other apps. Not a statement from the devs.
> > > If that truly is the case, could this solution be implemented as an unofficial companion app that adds the functionality?
> > 
> > 
> > I suspect they don't want to chase workarounds. At all. Once again, priorities.
> 
> Sorry, wasn't requesting a companion app from the Immich devs, more from Dreaming-Codes on if it was possible for this solution to be separated from the actual app. I'm not as familiar with Kotlin & Dart as I'd like to be at the moment so I was curious if that would be possible. I'm assuming it wouldn't be unless some sort of plugin functionality was implemented into Immich, which I know wouldn't be happening.
> 
> (wasn't requesting anything, just curious on what is possible)

If this new PR #26802 gets refused again I'll go ahead and open a new PR with only the document implementation and then make a separate app with the Cloud Media Provider Implementation.

**Reply by benjamin238** [reply]
It's really disappointing the immich team rejected this proposal; the lack of native photo picker integration is one of the biggest hurdles in using the software on android. Immich is fully self-hosted software that requires significant sysadmin knowledge to run - so I think it would be reasonable to expect user familiarity with ADB (or at least shizuku).

It would be a good idea to fork the immich app to provide this feature, given how well it boosts QoL.

**Reply by Dreaming-Codes** [reply]
> don't understand why we can't have a Shizuku toggle in the app's already existing advanced settings with a link to the Immich documentation on what ADB commands the toggle runs. Seems like a no-brainer. There is a solution for a major app inconvenience right here but we instead have to wait on Google to change their mind on this. Would love a mirror with these changes, this is the main thing keeping my family from us

Here you go @Kuuchuu 
https://github.com/Dreaming-Codes/immich-cloud-media

**Comment by NicholasFlamy** [comment]
Something that might at least partially cover the needs of people who want this is #1641 and at least that isn't being gatekept by Google.

**Comment by alextran1502** [comment]
There seems to be an active conversation in this thread regarding a potential implementation of the feature.

I just want to put this out there: we do want the feature; however, we want to balance maintainability, usability, and security. If the feature requires asking using to perform an ADB command, it opens up a lot of risk for nefarious guides online to target non-technical users who want to use the feature.

Thanks, and please consider the risk of promoting such implementation, which could impact others

**Reply by Kuuchuu** [reply]
Along with the idea the Shizuku could automate this meaning the user would not be manually entering any commands,

Requiring ADB does not meaningfully introduce new risk; it simply places the feature behind the same boundary Android already uses to distinguish advanced users from typical users. Enabling developer mode and executing ADB commands is already an intentional step that signals the user is operating outside the normal consumer workflow.

Malicious guides instructing users to run harmful ADB commands can already exist independent of any specific feature. Preventing legitimate functionality because such guides might exist does not reduce that risk; it only removes capability from users who are intentionally using advanced tooling.

A common compromise used in many projects is to clearly label the setup as advanced or unsupported and require manual configuration. That preserves maintainability while allowing capable users to enable the feature without affecting the standard user experience.

Given that Immich already targets a self-hosting audience accustomed to Docker, networking configuration, and other technical setup steps, an optional ADB-based configuration would fit naturally as an advanced feature without impacting typical users.

**Reply by NicholasFlamy** [reply]
> Given that Immich already targets a self-hosting audience accustomed to Docker

You'd be surprised how low that bar is.

**Comment by GGORG0** [comment]
I'd say that if we can't get the Cloud Media Provider approval, then why not just merge in the SAF/Document provider for now? Many apps still offer a "File" button that opens the file picker.

**Reply by griffi-gh** [reply]
tbf we could just merge both. 

the cloud media provider api is a pretty thin layer that just wraps the DocumentProvider and shouldn't require much maintenance or upkeep. 

and since Cloud Media Provider requires extensive testing to be approved by Google that would also allow power users to test it early if they choose so.

Even if cloud provider doesn't land, AT LEAST DocumentProvider should be implemented though. Its one of the main QOL hurdles with immich at the moment.

**Comment by Dreaming-Codes** [comment]
For anyone interested in a Cloud Media Provider implementation, I built a separate app that connects to your Immich instance and lets you use it as a Cloud Media Provider. Since the Immich team [explicitly stated](https://github.com/immich-app/immich/pull/26802#issuecomment-4069433601) that they do not want this feature in their codebase, I decided to implement it separately.

https://github.com/Dreaming-Codes/immich-cloud-media

Please note that this project is not affiliated with or endorsed by Immich.

**Comment by elio42** [comment]
## Proposal to have an OEM sponsor the Immich cloud media provider nomination

This idea was briefly mentioned by [Dreaming-Codes](https://github.com/Dreaming-Codes) but there was never any discussion around that:

> Hopefully, FUTU can get a smartphone manufacturer to add Immich to the allowed apps so we can avoid this hack.
>
> <cite>-- @Dreaming-Codes</cite>

### Situation

Reading through this Google page [here](https://developer.android.com/guide/topics/providers/cloud-media-provider), there only seem to be the following two conditions for nominations:

> (...) Only apps nominated by OEMs are eligible to participate in this program to become a cloud media provider for Android at this time. (...)

> (...) Nominated apps must meet all technical requirements and pass all quality tests. (...)

A variety of (very reasonable) technical requirements are listed at the bottom of that page.

**And then it says:**

> (...) Once approved, these apps become accessible as cloud media providers on any GMS Android-powered device on which they are installed.

### Idea:

I believe there are a number of OEMs that might be interested in nominating Immich as a cloud media provider for either or both of the following reasons:

- A service to their community (relevant especially for privacy focused brands)
- A great PR move because their brand would have enabled Immich to exist as a cloud media provider on Android

Looking into some brands I think there are four major candidates:

1. [Fairphone](https://shop.fairphone.com/home)
2. [Shift](https://www.shift.eco/)
3. [Teracube](https://myteracube.com/pages/teracube-2s)
4. [Nothing](https://nothing.tech/)

The first three suggestions are manufacturers that explicitly make fairly manufactured smartphones with some of them also offering de-googled versions.

Nothing isn't explicitly known for being privacy focused, but they are known for marketing stunts so they _might_ also be interested if none of the first three manufacturers are interested.

_And maybe there are more manufacturers that I don't know._

#### I think there are two major options for this to happen:

1. The Immich core team privately asks these companies for nominating Immich.
2. The Immich core team makes an open letter to privacy focused OEMs that have a GMS licence to nominate Immich. (With or without naming any company explicitly in that open letter)

### "But Google can just deny the request"

Yes. They explicitly say:

> Android is running a pilot program(...)

and with that, Google can reasonably deny any nomination without giving a good reason.

But this has been a "pilot program" (at least in name) for over 2 years now (started in Feb 2024). Calling a feature this important a "pilot" for over two years is... questionable at best. If Google gives a firm **"no"** on such a request it might at least make the situation move forward.

There is a chance that it will create _some_ media coverage. And there might also be fear of potential antitrust processes. Alphabet is officially a "Gatekeeper" company under the EU's DMA and Android is considered a "Core Platform Service". Specifically this _might_ (or might not) be considered "self-preferencing" and it _might_ fall under the DMA's interoperability obligations (Article 6.7), which require gatekeepers to give third-party apps the same access to OS features as their own services.

Either way, with a bit of luck this whole situation would at least move _somewhere_.

### Discussion

I’d love to hear the maintainers’ thoughts on whether this path could be considered to advance the situation. What are your thoughts on starting this conversation with a privacy focused manufacturer?

**Comment by SanjayDevTech** [comment]
Just to get myself on the same page - these OEM nominations only required for devices included with GMS or Playservices.
Will this apply to GrapheneOS or LineageOS with no Play services installed?

Even without Play services Google can change the allowlist in my phone?

**Reply by NicholasFlamy** [reply]
Pretty sure there's a whitelist that has to be configured at an OS level. Pretty sure it's AOSP and not GMS.

**Reply by elio42** [reply]
On [googles page](https://developer.android.com/guide/topics/providers/cloud-media-provider) it says the following:

> (...) Once approved, these apps become accessible as cloud media providers on any GMS Android-powered device on which they are installed.

So this means if Immich was approved as a cloud media provider, it would be accessible on **any** device that has google services.

Though I suppose that would probably also mean that it would not apply to OS' that don't use google services.

**Reply by Dreaming-Codes** [reply]
It is still a device config that Android checks, regardless of whether Google Play Services is installed. Google Play Services just acts as a device admin that changes that config, so even on a degoogled device that config still needs to be altered.

If you want this feature and are comfortable with running a one time ADB command to set it up, you can take a look at my app:
https://github.com/Dreaming-Codes/immich-cloud-media
If you find it useful, please consider starring the repo.

**Comment by kernelb00t** [comment]
Please note that if Google says no and the EU strikes back, we could even see that on iOS !
